### PR TITLE
IPMI driver Patch for AUTH 

### PR DESCRIPTION
**Pull request by SahilTikale**
Addresses merge conflicts pertaining to https://github.com/CCI-MOC/haas/pull/491
Included IPMI related code in the AUTH branch.

### CODE REVIEW

**Comment by SahilTikale** [comment]
Still working on `tests/unit/api/auth.py`

**Comment by SahilTikale** [comment]
@zenhack 
Can you suggest what would be the equivalent for
`node = model.Node(node_dict['label'], '', '', '')` to include the IPMI driver info ?

I am using the following for unit testing `model.py`

```
Node(label='node-99', obm=Ipmi(type="http://schema.massopencloud.org/haas/v0/obm/ipmi",
                  host= "ipmihost", user= "root", password= "tapeworm"))
```

File: docs/rest_api.md [hunk 1]
```diff
@@ -244,18 +244,28 @@ Possible Errors:
...
+
 `PUT /node/<node>`
 
-Register a node named `<node>` with the database.
+Request Body:
+        {"obm": { "type": <obm-subtype>,
+                <additional sub-type specific values>}
+        }
 
-Authorization requirements:
```
Comment (zenhack) [inline docs/rest_api.md:251]: Why was this removed?

File: haas/cli.py [hunk 2]
```diff
@@ -265,6 +265,37 @@ def headnode_stop(headnode):
...
+    else:
+        sys.stderr.write('ERROR: Wrong OBM subtype supplied\n')
+        sys.stderr.write('Supported OBM sub-types: ipmi, mock\n')
+        return
+
+    url = object_url('node', node)
+    do_put(url, data={"obm": obminfo})
+
 @cmd
 def node_register(node, ipmi_host, ipmi_user, ipmi_pass):
```
Comment (zenhack) [inline haas/cli.py:300]: You need to actually remove the old version of this function. Did you test this? I would expect the second definition to mask the first, making this change not actually do anything.

File: haas/test_common.py [hunk 3]
```diff
@@ -242,8 +242,14 @@ def site_layout():
         api.switch_register(**switch)
 
     for node in layout['nodes']:
-        api.node_register(node['name'], node['ipmi']['host'],
-            node['ipmi']['user'], node['ipmi']['pass'])
+        api.node_register(node['name'],
```
Comment (zenhack) [inline haas/test_common.py:245]: You shouldn't be passing in the individual fields from the obm object, just the whole thing, e.g. `api.node_register(node['name'], obm=node['obm']). This will prevent it from working with any other OBM driver that doesn't have exactly the same fields. Did I just miss this when reviewing your patch the first time, or did this go in as part of fixing the tests afterwards?

**Comment by zenhack** [comment]
@SahilTikale, the `model.Node(node_dict['label'], '', '', '')` is just calling Node's `__init__`; you should just be ble to make the same transformation as you made to everything else. It's just passing the empty string as user/host/pass, but you can set them to anything -- the tests are relying on `power_*` being decorated with `no_dry_run`. It might actually be better for them to be using the null or mock obm driver, but if that's going to take you a good amount of time I'd rather get this merged and file a cleanup bug for it; this has been sitting long enough, and I want to get it in before it bitrots again.

**Comment by SahilTikale** [comment]
Getting your PR merged is my top priority. Will do the necessary changes to the patch.

**Comment by zenhack** [comment]
I'm not sure I understand why you opened a second pr instead of updating this one, but okay, I'll close this.
